### PR TITLE
[DOCS] Add documentation for Painless field API

### PR DESCRIPTION
**Pull request by lockewritesdocs**
Adds a new section for accessing documents with the `field` API and explains the benefits of using this API over the old method of accessing document values.

Preview link: https://elasticsearch_83388.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/script-fields-api.html

Relates to #61388

### CODE REVIEW

File: docs/reference/scripting/using.asciidoc [hunk 1]
```diff
@@ -70,7 +70,7 @@ GET my-index-000001/_search
   "script_fields": {
     "my_doubled_field": {
       "script": { <1>
-        "source": "doc['my_field'].value * params['multiplier']", <2>
+        "source": "field('my_field').get(null) * params['multiplier']", <2>
```
Comment (stu-elastic) [inline docs/reference/scripting/using.asciidoc:73]: What do you think about using `0` as a default?

File: docs/reference/scripting/using.asciidoc [hunk 2]
```diff
@@ -197,6 +197,86 @@ them in other cases to remove ambiguity.
...
 
+[discrete]
+[[script-fields-api]]
+=== Access document fields
+The examples from previous sections use the `field` API to access document
+fields:
+
+[source,painless]
+----
+field('my_field').get(null)
```
Comment (stu-elastic) [inline docs/reference/scripting/using.asciidoc:208]: Let's talk about using the `$` syntax and handling different types.

File: docs/reference/scripting/using.asciidoc [hunk 3]
```diff
@@ -197,6 +197,86 @@ them in other cases to remove ambiguity.
...
+The `field` API fundamentally changes how you access documents in Painless.
+Previously, you had to declare `doc` and the field name that you wanted to
+access:
+
+[source,painless]
+----
+doc['my_field'].value
+----
+
+Accessing document fields this way didn't handle missing values, which meant
```
Comment (stu-elastic) [inline docs/reference/scripting/using.asciidoc:220]: Missing values or missing mappings.

File: docs/reference/scripting/using.asciidoc (hunk 1 above)
Comment (jdconrad) [reply docs/reference/scripting/using.asciidoc:73]: For primitive types (numbers and boolean), null cannot be used as the default, so it has to be some number -- @stu-elastic's suggestion of `0` makes sense to me.

File: docs/reference/scripting/using.asciidoc [hunk 4]
```diff
@@ -197,6 +197,86 @@ them in other cases to remove ambiguity.
...
+that your Painless scripts had to check that fields exist and create
+intermediary objects for handling such omissions.
+
+Instead, use the `field` API, which is the preferred approach to access
+documents in Painless. The `field` API abstracts access options (such as
+`_source` or `doc_values`) and provides a field-centric view of your documents.
+
+The `field` API always returns a non-null object that represents the given field,
```
Comment (jdconrad) [inline docs/reference/scripting/using.asciidoc:228]: I think we need to differentiate fields that return primitive types versus fields that return Object types. `non-null` isn't correct if the default is `null`.

**Review by jdconrad — COMMENTED** [review]
This is a solid start! Left a few pieces of general feedback.

A couple other thoughts:
* We should probably specify this is a feature preview and is subject to change somewhere.
* Some fields do not yet work with the fields API such as text or geo since we are still trying to figure out the best values there to return, so continued use of `doc` there is what is available.
* Need to differentiate between fields that return primitive types versus reference types since null can't be used for primitive defaults.

File: docs/reference/scripting/using.asciidoc (hunk 2 above)
Comment (stu-elastic) [inline docs/reference/scripting/using.asciidoc:208]: Let's use a different default than null here, empty string or 1.

File: docs/reference/scripting/using.asciidoc [hunk 5]
```diff
@@ -197,6 +197,86 @@ them in other cases to remove ambiguity.
...
+The examples from previous sections use the `field` API to access document
+fields:
+
+[source,painless]
+----
+field('my_field').get(null)
+----
+
+The `field` API fundamentally changes how you access documents in Painless.
+Previously, you had to declare `doc` and the field name that you wanted to
```
Comment (stu-elastic) [inline docs/reference/scripting/using.asciidoc:212]: TODO(stu/jack) write an explanation of `doc` as multi-valued, containing a key per mapping.

Comment (stu-elastic) [inline docs/reference/scripting/using.asciidoc:212]: ~declare `doc`~
access the `doc` map with the field name

File: docs/reference/scripting/using.asciidoc [hunk 6]
```diff
@@ -197,6 +197,86 @@ them in other cases to remove ambiguity.
...
+
+[discrete]
+==== Convenient, simpler access
+Instead of explicitly calling the `field` API with the `get()` method, you can
+include the `$` shortcut. Just include the `$` symbol, field name, and value
+that you want to retrieve:
+
+[source,painless]
+----
+$(‘field’, null)
```
Comment (stu-elastic) [inline docs/reference/scripting/using.asciidoc:245]: Perhaps `'default'` instead of a null?

File: docs/reference/scripting/using.asciidoc [hunk 7]
```diff
@@ -197,6 +197,86 @@ them in other cases to remove ambiguity.
...
+----
+
+With these enhanced capabilities and simplified syntax, you can write scripts
+that are shorter, less complex, and easier to parse. For example, the following
```
Comment (stu-elastic) [inline docs/reference/scripting/using.asciidoc:249]: ~easier to parse~
easier to read.

**Comment by stu-elastic** [comment]
Let's avoid converting things to `field` yet, and instead move the text to a new section one level up called "field access in painless".

File: docs/reference/scripting/using.asciidoc (hunk 1 above)
Comment (lockewritesdocs) [reply docs/reference/scripting/using.asciidoc:73]: I'll change `null` to `0` here

File: docs/reference/scripting/using.asciidoc (hunk 2 above)
Comment (lockewritesdocs) [reply docs/reference/scripting/using.asciidoc:208]: Since we're describing the basic syntax, let's just use an empty string for `get()`.

File: docs/reference/scripting/using.asciidoc (hunk 6 above)
Comment (lockewritesdocs) [reply docs/reference/scripting/using.asciidoc:245]: I'll change this to `<default>`

File: docs/reference/scripting/using.asciidoc (hunk 7 above)
Comment (lockewritesdocs) [reply docs/reference/scripting/using.asciidoc:249]: ➕ 1️⃣

File: docs/reference/scripting/access-fields.asciidoc [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+<titleabbrev>Access fields in a document</titleabbrev>
+++++
+
+beta::["The `field` API is still in development and should be considered a beta feature. The API is subject to change and this iteration is likely not the final state.",{es-issue}78920]
+
+Use the `field` API to access document fields:
+
+[source,painless]
+----
+field('my_field').get()
```
Comment (stu-elastic) [inline docs/reference/scripting/access-fields.asciidoc:13]: `field('my_field').get(<default value>)`

File: docs/reference/scripting/access-fields.asciidoc [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+you had to access the `doc` map with the field name that you wanted to access:
+
+[source,painless]
+----
+doc['my_field'].value
+----
+
+Accessing document fields this way didn't handle missing values or missing
+mappings, which meant that your Painless scripts had to check that fields exist
+and create intermediary objects for handling such omissions.
```
Comment (stu-elastic) [inline docs/reference/scripting/access-fields.asciidoc:26]: ~and create intermediary objects for handling such omissions.~

and that values exist in order write robust scripts.

File: docs/reference/scripting/access-fields.asciidoc [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+Instead, use the `field` API, which is the preferred approach to access
+documents in Painless. The `field` API abstracts access options (such as
+`_source` or `doc_values`) and provides a field-centric view of your documents.
```
Comment (stu-elastic) [inline docs/reference/scripting/access-fields.asciidoc:30]: ~The `field` API abstracts access options (such as
`_source` or `doc_values`) and provides a field-centric view of your documents.~

The `field` API makes it easy to handle missing values and, in the future, will abstract `_source` and `doc-value` access.

File: docs/reference/scripting/access-fields.asciidoc [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+NOTE: Some fields aren't yet compatible with the `fields` API, such as `text` or
+`geo` fields. since we are still trying to figure out the best values there to return, so continued use of doc there is what is available.
```
Comment (stu-elastic) [inline docs/reference/scripting/access-fields.asciidoc]: ~since we are still trying to figure out the best values there to return~
So continue to use doc for the few fields that are not supported in the new API.

File: docs/reference/scripting/access-fields.asciidoc [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+The `field` API returns a `Field` object that iterates over fields with 
+multiple values, providing access to the underlying value through the `get()`
```
Comment (stu-elastic) [inline docs/reference/scripting/access-fields.asciidoc:37]: `get(<default>)`

File: docs/reference/scripting/access-fields.asciidoc [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+The `field` API returns a `Field` object that iterates over fields with 
+multiple values, providing access to the underlying value through the `get()` 
+method, as well as type conversion and helper methods. 
+
+The `field` API returns the default value that you specify, regardless of
+whether the field exists or has any values for the current document.
+This means that the `field` API can handle missing values without requiring 
+additional logic. For a reference (`Object`) type such as `keyword`, the default 
+value can be `null`. For a primitive type such as `boolean` or `long`, the
+default value must be a non-null matching numerical type, such as `false` or `1`.
```
Comment (stu-elastic) [inline docs/reference/scripting/access-fields.asciidoc:45]: ~non-null matching numerical type~

matching primitive type

File: docs/reference/scripting/access-fields.asciidoc [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+This means that the `field` API can handle missing values without requiring 
+additional logic. For a reference (`Object`) type such as `keyword`, the default 
+value can be `null`. For a primitive type such as `boolean` or `long`, the
+default value must be a non-null matching numerical type, such as `false` or `1`. 
+
+[discrete]
+=== Convenient, simpler access
+Instead of explicitly calling the `field` API with the `get()` method, you can
+include the `$` shortcut. Just include the `$` symbol, field name, and value
+that you want to retrieve:
```
Comment (stu-elastic) [inline docs/reference/scripting/access-fields.asciidoc:51]: ~and value that you want to retrieve~

and a default value, in case the field does not have a value.

File: docs/reference/scripting/using.asciidoc [hunk 15]
```diff
@@ -168,7 +168,7 @@ GET my-index-000001/_search
   "script_fields": {
     "my_doubled_field": {
       "script": {
-        "source": "doc['my_field'].value * params['multiplier']",
+        "source": "field('my_field').get(null) * params['multiplier']",
```
Comment (stu-elastic) [inline docs/reference/scripting/using.asciidoc:171]: ~field('my_field').get(null)~

~field('my_field').get(1.0)~

File: docs/reference/scripting/access-fields.asciidoc (hunk 9 above)
Comment (lockewritesdocs) [reply docs/reference/scripting/access-fields.asciidoc:26]: To include your change, I'd like to rewrite this as:

>Accessing document fields this way didn't handle missing values or missing mappings, which meant that to write robust Painless scripts, you needed to include logic to check that both fields and values exist.

File: docs/reference/scripting/access-fields.asciidoc (hunk 10 above)
Comment (lockewritesdocs) [reply docs/reference/scripting/access-fields.asciidoc:30]: I'd like to rewrite this sentence to avoid "in the future", but also provide space for expansion in subsequent iterations:

>The `field` API handles missing values, and will evolve to abstract access to `_source` and `doc_values`.

File: docs/reference/scripting/access-fields.asciidoc (hunk 11 above)
Comment (lockewritesdocs) [reply docs/reference/scripting/access-fields.asciidoc]: I removed that last bit in a previous commit, and would like to rewrite this as:

>NOTE: Some fields aren't yet compatible with the `fields` API, such as `text` or `geo` fields. Continue using `doc` to access field types that the `field` API doesn't support.

File: docs/reference/scripting/access-fields.asciidoc (hunk 12 above)
Comment (lockewritesdocs) [reply docs/reference/scripting/access-fields.asciidoc:37]: I'll change this to `get(<default_value>)` to align with the change on [line 13](https://github.com/elastic/elasticsearch/pull/83388#discussion_r797840350).

File: docs/reference/scripting/access-fields.asciidoc [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+doesn't support.
+
+The `field` API returns a `Field` object that iterates over fields with 
+multiple values, providing access to the underlying value through the
+`get(<default_value>)` method, as well as type conversion and helper methods. 
+
+The `field` API returns the default value that you specify, regardless of
+whether the field exists or has any values for the current document.
+This means that the `field` API can handle missing values without requiring 
+additional logic. For a reference (`Object`) type such as `keyword`, the default
```
Comment (stu-elastic) [inline docs/reference/scripting/access-fields.asciidoc:43]: Lets remove (`Object`) here because there is an Object type and that could be confusing.

**Review by stu-elastic — APPROVED** [review]
Very nice work

File: docs/reference/scripting/access-fields.asciidoc [hunk 17]
```diff
@@ -84,4 +84,42 @@ operating on them:
...
+`field` API (from the `get` and `as<Type>` methods) and the `doc` map (from the
+`getValue` and `get` methods). 
+
+NOTE: Some fields are not currently supported. For the most current list of
+supported fields, refer to {es-issue}79105[#79105].
+
+[cols="1,1,1",options="header",]
+|========
+|Mapped field type
+|Returned type from `field` (`get`,`asType`)
```
Comment (jdconrad) [inline docs/reference/scripting/access-fields.asciidoc:102]: asType becomes something like asBigInteger or asString which I'm not sure is clear from this. Is there a way to show that Type -> BigInteger, etc.

File: docs/reference/scripting/access-fields.asciidoc [hunk 18]
```diff
@@ -84,4 +84,42 @@ operating on them:
...
+|Returned type from `doc` (`getValue`,`get`)
+  |`binary`           |`ByteBuffer`               |`BytesRef`
+  |`boolean`          |`boolean`                  |`boolean/Boolean`
```
Comment (jdconrad) [inline docs/reference/scripting/access-fields.asciidoc:105]: Should boolean/Boolean become boolean, Boolean here?

File: docs/reference/scripting/access-fields.asciidoc [hunk 19]
```diff
@@ -84,4 +84,42 @@ operating on them:
...
+----
+
+[discrete]
+=== Supported mapped field types
+The following table indicates the mapped field types that the `field` API
+supports. For each supported type, values are listed that are returned by the
+`field` API (from the `get` and `as<Type>` methods) and the `doc` map (from the
+`getValue` and `get` methods). 
+
+NOTE: Some fields are not currently supported. For the most current list of
```
Comment (jdconrad) [inline docs/reference/scripting/access-fields.asciidoc:96]: Did we want to say unsupported fields are still accessible through `doc`?

**Review by jdconrad — APPROVED** [review]
This looks pretty good! Left a few comments to be addressed still.

File: docs/reference/scripting/access-fields.asciidoc (hunk 19 above)
Comment (lockewritesdocs) [reply docs/reference/scripting/access-fields.asciidoc:96]: Yeah, I think that's a good addition. I'm thinking that we can rephrase this as:

>NOTE: The `fields` API currently doesn't support some fields, but you can still access those fields through the `doc` map. For the most current list of supported fields, refer to {es-issue}79105[#79105].

**Review by jdconrad — APPROVED** [review]
Table looks good to me! Thank you!
